src/nvim/ops.c

```diff
@@ -790,61 +778,45 @@ void get_yank_register(int regname, int mode)
     i = CharOrdLow(regname) + 10;
   else if (ASCII_ISUPPER(regname)) {
     i = CharOrdUp(regname) + 10;
-    y_append = true;
   } else if (regname == '-')
     i = DELETION_REGISTER;
   else if (regname == '*')
     i = STAR_REGISTER;
   else if (regname == '+')
     i = PLUS_REGISTER;
-  y_current = &(y_regs[i]);
+  reg = &y_regs[i];
 
   if (mode == YREG_YANK) {
     // remember the written register for unnamed paste
-    y_previous = y_current;
+    y_previous = reg;
   }
+  return reg;
 }
 
-
-/*
- * Obtain the contents of a "normal" register. The register is made empty.
- * The returned pointer has allocated memory, use put_register() later.
- */
-void *
-get_register (
-    int name,
-    int copy               /* make a copy, if FALSE make register empty. */
-) FUNC_ATTR_NONNULL_RET
+static bool is_append_register(int regname)
 {
-  get_yank_register(name, YREG_PASTE);
-
-  struct yankreg *reg = xmalloc(sizeof(struct yankreg));
-  *reg = *y_current;
-  if (copy) {
-    if (reg->y_size == 0) {
-      reg->y_array = NULL;
-    } else {
-      reg->y_array = xmalloc(reg->y_size * sizeof(char_u *));
-      for (linenr_T i = 0; i < reg->y_size; ++i) {
-        reg->y_array[i] = vim_strsave(y_current->y_array[i]);
-      }
-    }
-  } else
-    y_current->y_array = NULL;
-
-  return (void *)reg;
+  return ASCII_ISUPPER(regname);
 }
 
-/*
- * Put "reg" into register "name".  Free any previous contents and "reg".
- */
-void put_register(int name, void *reg)
+/// Returns a copy of contents in register `name`
+/// for use in do_put. Should be freed by caller.
+yankreg_T *copy_register(int name)
+  FUNC_ATTR_MALLOC
+  FUNC_ATTR_NONNULL_RET
 {
-  get_yank_register(name, YREG_PUT);
-  free_yank_all();
-  *y_current = *(struct yankreg *)reg;
-  xfree(reg);
-  set_clipboard(name);
+  yankreg_T *reg = get_yank_register(name, YREG_PASTE);
+
+  yankreg_T *copy = xmalloc(sizeof(yankreg_T));
+  *copy = *reg;
+  if (copy->y_size == 0) {
+    copy->y_array = NULL;
+  } else {
+    copy->y_array = xmalloc(copy->y_size * sizeof(char_u *));
```

Comment:
Could the multiplication overflow if the register contains a huge amount of text? I'm asking because I haven't checked elsewhere.